### PR TITLE
Add tensor clone builtin (#38437)

### PR DESCRIPTION
**Pull request by wconstab**
- add corresponding test
 - fix prior lint errors in test_jit.py

### CODE REVIEW

**Comment by wconstab** [comment]
@eellison directional feedback would be great, as I spent most of the time understanding the flow of the code and just threw the fix/test together.  I'm still trying to figure out if i broke something else in test_jit or my local env is hosed.

File: test/test_jit.py [hunk 1]
```diff
@@ -17192,6 +17196,12 @@ def fn(x=None):
         self.checkScript(fn, ((3, 4),))
         self.checkScript(fn, ())
 
+    def test_tensor_clone(self):
```
Comment (eellison) [inline test/test_jit.py:17199]: can you add the test here to https://github.com/pytorch/pytorch/blob/master/test/test_jit.py#L8107 ? I think you just need to add some tensor constructors to `lists`. you may need to only add them when we're testing `tensor` and not `as_tensor`, and you may need to invoke `with freeze_rng_state():` if you're using rand initializers.

File: torch/csrc/jit/frontend/builtin_functions.cpp [hunk 2]
```diff
@@ -72,6 +72,12 @@ def _test_serialization_subcmul_0_2(self: Tensor, other:Tensor, alpha: number=2)
   return other - (self * alpha)
 )SCRIPT";
 
+// Tensor copy creation
+auto tensor_clone = R"SCRIPT(
+def tensor(other:Tensor) -> Tensor:
```
Comment (eellison) [inline torch/csrc/jit/frontend/builtin_functions.cpp:77]: I think we also need to handle the other arguments:
https://pytorch.org/docs/stable/torch.html#torch.tensor `torch.tensor(data, dtype=None, device=None, requires_grad=False, pin_memory=False)`.
```
if (dtype is not None) {
    tensor = tensor.to(dtype)
}
...
```

**Review by eellison — COMMENTED** [review]
> I'm still trying to figure out if i broke something else in test_jit or my local env is hosed. 

What do you mean exactly ?

And yea, makes sense! The flow of compilation is not simple. Feel free to ping me if you have any questions.
This looks good! I left a few comments. The general idea is we have a core set of ops, and desugar higher level ops to the core set. We haven't always done a great job of keeping the op set small but hopefully will be doing a better job moving forward.

I actually realized one thing, which is that right now builtin_functions gets compiled and inlined eagerly, which breaks when builtin_functions.cpp have control flow 😢 So, I think we might have to shelve this PR until that gets fixed.

**Comment by wconstab** [comment]
i think i follow your suggestions, i'll spend some more time on it to understand fully - and hopefully also get to the point where I see the issue you're describing RE control flow.  I presume you're saying that i need to add CF e.g. to support the optional dtype in tensor creation, and that this part wouldn't work due to the way builtins are compiled currently.

Tangentially, it seems like I did break something so I'm trying to figure that out too.  

Is the builtins compile fix you mentioned something that's already in the queue/planned?

File: torch/csrc/jit/frontend/builtin_functions.cpp (hunk 2 above)
Comment (wconstab) [reply torch/csrc/jit/frontend/builtin_functions.cpp:77]: my assumption was that tensor.clone() would result in a new tensor with the same properties as the original tensor (e.g. dtype etc.) -- is this not the case? 

are you concerned with copying the dtype from the src tensor, or with providing a convenience wrapper to override it?

oh, i think i get it now- the use case is something like this, i guess?
`@script
def foo(x):
   return torch.tensor(x, dtype=blah)`

Comment (eellison) [reply torch/csrc/jit/frontend/builtin_functions.cpp:77]: yep, correct

**Comment by eellison** [comment]
@wconstab i can try to get to it soon. it's not being actively being worked on. it's probably not a great intro task bc bulitin_functions is sort of a special cased part of the system and trying to serialize it will probably run into a few weird issues.

**Comment by wconstab** [comment]
ok, no rush just wondered. i'll push the test change now and fool around with the other part a little but probably move to a new task.

File: test/test_jit.py (hunk 1 above)
Comment (wconstab) [reply test/test_jit.py]: ok this part should be done in the latest commit.

**Comment by wconstab** [comment]
@eellison - following up on
> I actually realized one thing, which is that right now builtin_functions gets compiled and inlined eagerly, which breaks when builtin_functions.cpp have control flow 😢 So, I think we might have to shelve this PR until that gets fixed.

Has there been any progress on fixing how builtin_functions gets compiled so that it works with control flow?  If not, is there an issue filed to track that, which I could link this issue to?

**Comment by eellison** [comment]
@wconstab no there's not. I figured it would come up sooner or later but it hasn't yet. There's one sort of onboarding-hack a month type proposal I have written up that would also require this FWIW.

**Pull request closed without merging**